### PR TITLE
fix: avoid duplicate request of meeting info by passing it from cantina

### PR DESCRIPTION
**Pull request by maxinteger**
# COMPLETES SPARK-489305

## This pull request addresses

Avoid duplicate request of meeting info

## by making the following changes

Adding extra parameter to `meeting.create` method to pass down pre-fetched meeting info, also adding utility methods to process the new parameter.

### Change Type

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested

< ENUMERATE TESTS PERFORMED, WHETHER MANUAL OR AUTOMATED >

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [ ] I discussed changes with code owners prior to submitting this pull request

- [ ] I have not skipped any automated checks
- [ ] All existing and new tests passed
- [ ] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 1]
```diff
@@ -1386,6 +1393,74 @@ export default class Meeting extends StatelessWebexPlugin {
...
+        file: 'meetings',
+        function: 'fetchMeetingInfo',
+      },
+      EVENT_TRIGGERS.MEETING_INFO_AVAILABLE
+    );
+
+    this.updateMeetingActions();
+  }
+
+  public async injectMeetingInfo(info: any, fetchParams: FetchMeetingInfoParams): Promise<void> {
```
Comment (martroos) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:1429]: nitpick: The linter complains about this lacking a JSDoc comment.

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 2]
```diff
@@ -1386,6 +1393,74 @@ export default class Meeting extends StatelessWebexPlugin {
...
+    await this.prepForFetchMeetingInfo(fetchParams, 'injectMeetingInfo');
+
+    this.setMeetingInfo(info, undefined);
+  }
+
+  private prepForFetchMeetingInfo(
```
Comment (martroos) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:1435]: nitpick: The linter complains about this lacking a JSDoc comment.

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js
Comment (martroos) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js:1]: nitpick: Would be much easier to review this if it didn't contain all the auto-formatter changes (or at least if they were done in a separate commit).

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js [hunk 3]
```diff
@@ -4621,6 +4786,105 @@ describe('plugin-meetings', () => {
...
+          conversationUrl: 'some_convo_url',
+          locusUrl: 'some_locus_url',
+          sipUrl: 'some_sip_url', // or sipMeetingUri
+          meetingNumber: '123456', // this.config.experimental.enableUnifiedMeetings
+          hostId: 'some_host_id', // this.owner;
+        };
+        const FAKE_MEETING_INFO_LOOKUP_URL = 'meetingLookupUrl';
+
+        it('calls meetingInfoProvider with all the right parameters and parses the result', async () => {
+          meeting.requiredCaptcha = FAKE_SDK_CAPTCHA_INFO;
```
Comment (martroos) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js:4811]: suggestion: This is a little confusing, since for the inject case, a captcha is never needed - it is initialised as `null` and is only set to something non-`null` when doing an actual fetch. I realise the reason for this is because you also provide a captcha code, but again I don't think this should be provided generally for inject info. I think it may be simpler to just not set / provide anything captcha (or password) related for this first test.

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js [hunk 4]
```diff
@@ -4621,6 +4786,105 @@ describe('plugin-meetings', () => {
...
+            meeting,
+            {file: 'meetings', function: 'fetchMeetingInfo'},
+            'meeting:meetingInfoAvailable'
+          );
+          assert.calledWith(meeting.updateMeetingActions);
+        });
+
+        it('fails if captchaCode is provided when captcha not needed', async () => {
+          meeting.attrs.meetingInfoProvider = {
+            injectMeetingInfo: sinon.stub().resolves(),
```
Comment (martroos) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js:4845]: question: Why is this necessary? `injectMeetingInfo` isn't a property of `meetingInfoProvider` (so of course it doesn't get called). Should this actually be `fetchMeetingInfo`?

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js [hunk 5]
```diff
@@ -4621,6 +4786,105 @@ describe('plugin-meetings', () => {
...
+            Error,
+            'injectMeetingInfo() called with captchaCode when captcha was not required'
+          );
+
+          assert.notCalled(meeting.attrs.meetingInfoProvider.injectMeetingInfo);
+        });
+
+        it('fails if password is provided when not required', async () => {
+          meeting.attrs.meetingInfoProvider = {
+            injectMeetingInfo: sinon.stub().resolves(),
```
Comment (martroos) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js:4867]: question: Same question as before regarding if this should actually be `fetchMeetingInfo`

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js [hunk 6]
```diff
@@ -4621,6 +4786,105 @@ describe('plugin-meetings', () => {
         });
       });
 
+      describe('#injectMeetingInfo', () => {
```
Comment (martroos) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js:4789]: suggestion: Theoretically there could be a delayed fetch pending and `injectMeetingInfo` could be called during this time, so it might be worth checking that `fetchMeetingInfoTimeoutId` gets cleared when this is the case.

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js [hunk 7]
```diff
@@ -4621,6 +4786,105 @@ describe('plugin-meetings', () => {
...
+          assert.calledWith(
+            TriggerProxy.trigger,
+            meeting,
+            {file: 'meetings', function: 'fetchMeetingInfo'},
+            'meeting:meetingInfoAvailable'
+          );
+          assert.calledWith(meeting.updateMeetingActions);
+        });
+
+        it('fails if captchaCode is provided when captcha not needed', async () => {
```
Comment (martroos) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js:4843]: suggestion: Would be nice to check that the `meetingInfoAvailable` event isn't triggered in this (and the password) case.

**Review by martroos — CHANGES_REQUESTED** [review]
Requesting changes mostly to have questions answered before approving. I'm a little concerned that this will only work with complete meeting info (not partial info as you might get from a fetch when a password / captcha is needed but not provided), but documentation should suffice for now (unless there is a reliable way of checking complete info is provided).

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 8]
```diff
@@ -1386,6 +1393,74 @@ export default class Meeting extends StatelessWebexPlugin {
...
+      EVENT_TRIGGERS.MEETING_INFO_AVAILABLE
+    );
+
+    this.updateMeetingActions();
+  }
+
+  public async injectMeetingInfo(info: any, fetchParams: FetchMeetingInfoParams): Promise<void> {
+    await this.prepForFetchMeetingInfo(fetchParams, 'injectMeetingInfo');
+
+    this.setMeetingInfo(info, undefined);
```
[the remark this answers is not in the document]
Comment (maxinteger) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:1432]: I updated the JSDoc and set the default value for the `meetingLookupUrl` parameter to `undefined`

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 9]
```diff
@@ -3214,7 +3234,7 @@ export default class Meeting extends StatelessWebexPlugin {
       | any,
     destination: object | string | null = null
   ) {
-    const webexMeetingInfo = meetingInfo?.body;
+    const webexMeetingInfo = meetingInfo?.body || meetingInfo;
```
[the remark this answers is not in the document]
Comment (maxinteger) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:3237]: We needed the `meetingInfo.body` format because later it uses `meetingInfo.error`

I added an error parameter and moved `parseMeetingInfo` out from `setMeetingInfo` to make it more clear

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js
Comment (maxinteger) [reply packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js:1]: Sorry about that. I was lazy to undo my commit. Next time I will separate them

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 10]
```diff
@@ -1386,6 +1393,74 @@ export default class Meeting extends StatelessWebexPlugin {
...
+  private setMeetingInfo(info, meetingLookupUrl) {
+    this.requiredCaptcha = null;
+    if (
+      this.passwordStatus === PASSWORD_STATUS.REQUIRED ||
+      this.passwordStatus === PASSWORD_STATUS.VERIFIED
+    ) {
+      this.passwordStatus = PASSWORD_STATUS.VERIFIED;
+    } else {
+      this.passwordStatus = PASSWORD_STATUS.NOT_REQUIRED;
+    }
```
[the remark this answers is not in the document]
Comment (maxinteger) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:1414]: I moved this and the `parseMeetingInfo` independently and I forget to restore the original order.

I moved out `parseMeetingInfo` from the `setMeetingInfo` and no the order is the same as in the original code.

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (martroos) [reply packages/@webex/plugin-meetings/src/meeting/index.ts]: My point wasn't that we shouldn't be providing `undefined` as an argument, but that we should allow for the lookup URL to be provided when calling `injectMeetingInfo`, so that externally we can state where the information came from when injecting it. For example, in guest join cases, it will come from a preJoin API. Knowing this helps us debug issues with meeting info. So my suggestion was effectively that we should provide an extra argument / add to options to provide this here and in the create method (where this is used). i.e.

```js
public async injectMeetingInfo(info: any, fetchParams: FetchMeetingInfoParams, meetingLookupUrl: string) {
```

Or something like that.

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js (hunk 7 above)
Comment (martroos) [reply packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js:4835]: The checks for the event triggers appear to be commented out - did they not work or was this just work in progress?

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (maxinteger) [reply packages/@webex/plugin-meetings/src/meeting/index.ts]: I see you point now. 

I refactored it, so meetingLookUp url can passed from cantina if it is available

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js (hunk 7 above)
Comment (maxinteger) [reply packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js:4835]: Good catch

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 11]
```diff
@@ -1432,16 +1432,18 @@ export default class Meeting extends StatelessWebexPlugin {
...
    * @param {Object} meetingInfo - Complete meeting info
    * @param {FetchMeetingInfoParams} fetchParams - Fetch parameters for validation
+   * @param {String} meetingLookupUrl - Lookup url, defined when the meeting info fetched
    * @returns {Promise<void>}
    */
   public async injectMeetingInfo(
     meetingInfo: any,
-    fetchParams: FetchMeetingInfoParams
+    fetchParams: FetchMeetingInfoParams,
+    meetingLookupUrl: string
```
Comment (martroos) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:1441]: suggestion: This additional parameter needs unit test(s) adding.

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 12]
```diff
@@ -1432,16 +1432,18 @@ export default class Meeting extends StatelessWebexPlugin {
    *
    * @param {Object} meetingInfo - Complete meeting info
    * @param {FetchMeetingInfoParams} fetchParams - Fetch parameters for validation
+   * @param {String} meetingLookupUrl - Lookup url, defined when the meeting info fetched
```
Comment (martroos) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:1435]: nitpick: This still may be undefined

**Review by martroos — APPROVED** [review]
